File: python/collections.py
[x] = (1, ) # get element from one element tuple, we use list here
(x, ) = (1, ) # the same



# SLICES
# You can create slice namually 
# And pass them into collection
t = ('Tom', 'Hardy', 1980, 2, 3)

NAME = slice(2) # [:2]
YEAR_MONTH = slice(2, 4) # [2:4]
DATE = slice(2, None)

print(t[NAME]) # ('Tom', 'Hardy')
print(t[YEAR_MONTH]) # (1980, 2)
print(t[DATE]) # (1980, 2, 3)


# REVERSED 
t1 = tuple(reversed((1, 2, 3))) # (3, 2, 1)
t2 = (1, 2, 3)[::-1] # the same
l = list(reversed(range(1, 4))) # [3, 2, 1]



# Concatenate tuples
t1, t2 = (1, 2), (3, 4)
print(t1 + t2) # (1, 2, 3, 4)
# Compare tuples lexicographically
print((1, 2) < (1, 3)) # True
print((3, 1) < (2, 3)) # False




# NAMEDTUPLE
# Namedtuple is like a class factory.
from collections import namedtuple

Person = namedtuple('Person', ['name', 'age'])
person = Person(name='Tom', age=25)
print(person.name)
print(person.age)

# Note we use underscope here, to have an ability to create attributes fiels, asdict
print(person._fields) # ('name', 'age')
print(person._asdict()) # {'name': 'Tom', 'age': 25}

another_person = person._replace(name='Jack') # create new namedtuple and change some fields
print(another_person.name) # 'Jack' 




# LISTS
# Multiplication just copies links to the inner elements n times
l1 = [[0], [1]]
l2 = l1 * 2 # [[0], [1], [0], [1]]
l2[0][0] = 3 # [[3], [1], [3], [1]]

# It's better to use generators to create 2-dimentional arrays
arr = [[0] * 5 for i in range(5)] # array 5 * 5


# CHANGING
x = [1, 2, 3]
x.append(4)
print(x) # [1, 2, 3, 4]
x.extend([7, 8]) # pass something iterable
print(x) # [1, 2, 3, 4, 7, 8]

x[1:3] = [9, 9, 9] # change several elements
print(x) # [1, 9, 9, 9, 4, 7, 8] 

x = [1, 2]
x.insert(1, 3) 
print(x) # [1, 3, 2]
x.insert(-1, 4)
print(x) # [1, 3, 4, 2]

l1 = [1, 2]
l2 = [3, 4]
l3 = l1 + l2
print(l3) # [1, 2, 3, 4]
l2 += l1 # inline appending, approximately equal l2.extend(l1)
print(l2) # [3, 4, 1, 2] 

# Delete elements
x = [1, 2, 3, 4]
del x[1]
print(x) # [1, 3, 4]
del x[1:] # delete several elements
print(x) # [1] 

# remove element and return it
x = [1, 2, 3, 4]
y = x.pop(2)
print(x) # [1, 2, 4]
print(y) # 3

# remove by value
x = [1, 2, 3, 4]
x.remove(3) 
print(x) # [1, 2, 4]

#reverse
x = [1, 2, 3]
x.reverse()
print(x) # [3, 2, 1]

# sort
x = [3, 2, 1]
res = sorted(x) # return sorted list
print(res) # [1, 2, 3]
x.sort() # inplace sorting
print(x) # [1, 2, 3]

x = [3, 2, 1]
x.sort(key=lambda x: x % 2, reverse=True) # specify key, and by ascending or descending

# cmp_to_key
# you can create comparator and then pass it into cmp_to_key to create key for sorting
from functools import cmp_to_key

x = [(1, 1), (2, 1), (2, 2), (1, 2)]
# sort by first ascending, second descending
def cmp(a, b):
    if(a[0] < b[0]):
        return -1
    if(a[0] > b[0]):
        return 1
    if(a[1] > b[1]):
        return -1
    if(a[1] < b[1]):
        return 1
    return 0

x.sort(key=cmp_to_key(cmp)) 
print(x) # [(1, 2), (1, 1), (2, 2), (2, 1)]




# DEQUE
# For list l.pop(0) works for linear times
# Use deque for constant times
# Access by index for linear time in this case
from collections import deque

d = deque([1, 2, 3])
print(d)

d.appendleft(1)
d.append(3)
print(d) # deque([1, 1, 2, 3, 3])

d.pop()
d.popleft()
print(d) # deque([1, 2, 3])

# You can specify max deque length, in this case excess elements will be automatically deleted.
d = deque([1, 2], maxlen=2)
d.append(3) 
print(d) # deque([2, 3], maxlen=2)

d.appendleft(4)
print(d) # deque([4, 2], maxlen=2)




# SET
s1, s2, s3 = {1, 2}, {2, 3}, {3, 4}
print(set.intersection(s1, s2, s3)) # {}
print(s1 & s2 & s3) # {}

print(set.union(s1, s2, s3)) # {1, 2, 3, 4}
print(s1 | s2 | s3) # {1, 2, 3, 4}

print(set.difference(s1, s2, s3)) # {1}
print(s1 - s2 - s3) # {1}

s1, s2 = {1, 2}, {2, 3}
print({1, 2}.isdisjoint({3, 4})) # is intersection empty
print({1, 2} <= {1, 2}) # is subset
print({1, 2} < {1, 2}) # is strict subset

# Add elements
s = set()
s.add(1) # add one element
s.update([2, 3]) # add several elements

s = set()
s.update([1, 2], [4, 5]) # {1, 2, 4, 5} you can pass several iterable objects 

# Remove elements
s = set([1, 2, 3])
s.remove(5) # Error! throws the KeyError if the set has no specified element
s.discard(5) # ignore, if the set has no specified element
s.clear() # clear all elements

# Set only can contains hashable elements, for collections only unchangeble
s = {[], []} # error, as list is changable
s = {}

# Unchangable set - frozenset
s = { frozenset([1, 2]), frozenset([3, 4]) } # correct




# DICTIONARY
# Create
d1 = dict(a=1, b=2) # with keyword arguments { 'a': 1, 'b':2 }
d2 = dict(d1) # with another dict { 'a': 1, 'b':2 }
d3 = dict(d1, c=3) # mixed { 'a': 1, 'b': 2, 'c': 3 }

# fromkeys, it takes enumerable collection of keys
d1 = dict.fromkeys(['a', 'b']) # { 'a': None, 'b': None }
d2 = dict.fromkeys(['a', 'b'], 0) # specify values for keys { 'a': 0, 'b': 0 }

# fromkeys uses default value for every key
d1 = dict.fromkeys([1, 2], [])
d1[1].append(2)
print(d1[2]) # [2]

#keys, values, items - projections of a dictionary
d = { 'a': 1, 'b': 2 }
print(d.keys()) # dict_keys(['a', 'b'])
print(d.values()) # dict_values([1, 2])
print(d.items()) # dict_items([('a', 1), ('b', 2)])

print(len(d.keys()))
print('a' in d.keys())

# Iteration
# You can iterate over key
d = { 'a': 1, 'b': 2 }
for k in d:
    print(d[k])
    
# You can't change collection while iteration
d = { 'a': 1, 'b': 2 }
for k in d:
    del d[k] # Error

# You should create another collection instead:
for k in set(d):
    del d[k] # Correct
    
# if you try to get a value for unknown key, there will be an error.
d = { 'a': 1, 'b': 2 }
print(d['c']) # error
# You can specify default value for this case with the method get
d = { 'a': 1, 'b': 2 }
print(d.get('c', 3)) # 3

# You can create element in dict, if there is no this key
d = { 'a': 1 }
print(d.setdefault('a', 2)) # 1  d = { 'a': 1 }
print(d.setdefault('b', 2)) # 2  d = { 'a': 1, 'b': 2 }

# You can add or update values in a dict with the update method
# Pass key-value as collection of tuples or a keyword argument
d = { 'a': 1, 'b': 2 }
d.update([('a', 4), ('b', 5)], c=6) # { 'a': 4, 'b': 5, 'c': 6 }

# You can delete elements with del and pop. Pop returns the deleted value.
d = { 'a': 1, 'b': 2 }
del d['a']
print(d.pop('b')) # 2
print(d) # {}



# DEFAULTDICT
# If we try to get element, that is not in dict there will be an errer.
# defaultdict create element in this case
# you should pass function that return the default value into constructor
from collections import defaultdict

d = defaultdict(lambda: 0, [('a', 1), ('b', 2)]) # default value is 0 
print(d['c']) # d = { 'a': 1, 'b': 2, 'c': 0 }



# ORDEREDDICT
# ordereddict enumerate elements in order of them adding into collection
# updating elements in collection doesn't change the order
from collections import OrderedDict

d = OrderedDict([('a', 1), ('b', 2)])
for k in d:
    print(k) # a b

d['c'] = 3
for k in d:
    print(k) # a b c
    
    
    
# COUNTER
# Counter count items occurance
from collections import Counter

counter = Counter({'a': 1, 'b': 2}) # {'a': 1, 'b': 2} 
counter = Counter(['a', 'b', 'a']) # {'a': 1, 'b': 2}
print(counter.most_common(1)) # [('b': 2)] return several most common elements
print(list(counter.elements())) # ['a', 'a', 'b'] get element 


counter['b'] += 4 # update value
counter.pop('a') # delete and return 

# It's also a default dict. Counter create element with 0 value
print(counter['c']) # 0

# Add values
counter.update(['a', 'b'])
counter.update({'a': 2, 'b': 3})

# Subtract values
counter.subtract(['a', 'b'])
counter.subtract({'a': 2, 'b': 3})

# Union operations
# Result counter contains only positive values
from collections import Counter

counter1 = Counter({'a': 1, 'b': 2})
counter2 = Counter({'a': 3, 'b': 1})

print(counter1 + counter2) # sum Counter({'a': 4, 'b': 3})
print(counter1 - counter2) # subtract Counter({'b': 1})
print(counter1 | counter2) # max Counter({'a': 3, 'b': 2})
print(counter1 & counter2) # min Counter({'a': 1, 'b': 1})

File: java/collections.java
// ITERABLE<E> AND ITERATOR<E> 
// The base interface Iterable<E>
public interface Iterable<E> {
    Iterator<E> iterator(); // returns iterator
    //Some another methods
}
// The base interface Iterator<E>
public interface Iterator<E> {
    boolean hasNext();
	E next();
	default void remove() {
        throw new UnsupportedOperationException("remove");
    }
}

// Example
static void printAll(Iterable<?> iterable) {
	Iterator<?> iterator = iterable.iterator();
	while(iterator.hasNext()) {
		Object value = iterator.next();
		System.out.println(value);
	}
}
// Or with for, it is converted in above on bitecode level
static void printAll(Iterable<?> iterable) {
	for(Object object: iterable) {
		System.out.println(object);
	}
}

// Iterator has method remove to remove current element from iterable
// It can throw UnsupportedOperationException, if collection is unchangable
public static void main(String[] args) throws IOException {
	ArrayList<String> strings = new ArrayList<>() {{
		add("aa"); add(""); add("bb"); add("");
	}};
	removeEmpty(strings); // remove empty strings
}
static void removeEmpty(Iterable<String> strings) {
	Iterator<String> iterator = strings.iterator();
	while(iterator.hasNext()) {
		String string = iterator.next();
		if(string.isEmpty()) iterator.remove();
	}
}

// Example, create collection with n the same elements
static <T> Iterable<T> nCopies(T value, int count) {
	if(count < 0) {
		throw new IllegalArgumentException("Count < " + 0);
	}
	return new Iterable<T>() { // we use anonymous class
		@Override
		public Iterator<T> iterator() { // we use anonymous class
			return new Iterator<T>() {
				int rest = count;

				@Override
				public boolean hasNext() { return rest > 0; }

				@Override
				public T next() {
					if(rest == 0) {
						throw new UnsupportedOperationException();
					}
					rest--;
					return value;
				}
			};
		}
	};
}



// COLLECTION<E>
public interface Collection<E> extends Iterable<E> {
	int size();
	boolean isEmpty();
	boolean contains(Object o); // accepts element of any type
    boolean containsAll(Collection<?> c); // collection elements of any type
    Object[] toArray();
    <T> T[] toArray(T[] a);

	// Methods that modifies collection, if collection readonly
	// they throw UnsupporteOperationException()
    boolean add(E e);
    boolean remove(Object o); // try to remove element of any type
    boolean addAll(Collection<? extends E> c);
    boolean removeAll(Collection<?> c); // try to remove element from collection of any type
    boolean retainAll(Collection<?> c); // save only elements from collection c
    void clear();	
}



// SET<E>
// Set<E> extends Collection<E>
// There is no new methods.
// Collection of different objects.
// To sets are equal if they contain the same elements.
// It's your problem if you change element that is Set
// Example, rangeSet:
public static Set<Integer> rangeSet(int fromInclusive, int toExclusive) {
	return new AbstractSet<Integer>() {
		@Override
		public Iterator<Integer> iterator() {
			return new Iterator<Integer>() {
				int current = fromInclusive;

				@Override
				public boolean hasNext() {
					return current < toExclusive;
				}
				@Override
				public Integer next() {
					if(current == toExclusive) {
						throw new IllegalArgumentException();
					}
					return current++;
				}
			};
		};

		@Override
		public int size() {
			return toExclusive - fromInclusive;
		}
	};
}

// Standard sets
// HashSet - standard changable unordered set
// LinkedHashSet - standard changable ordered set in order of adding
// TreeSet - changable sorted set(you can specify sorting method)
// EnumSet - changable set of enum. You can store several value of the enum
// Collections.emptySet() - unchangable empty set
// Collections.singleton(x) - unchangable set of one element
// Set.of(...) - unchangable unordered set of passed elements(you can passed
// several elements)
// Set.copyOf(...) - unchangable copy
// Collections.unmodifieableSet(set) - unchangable wrapper under the set
// Collections.synchronizedSet(set) - syncronized wrapper under the set
// Collections.checkedSet(set, type) - checkable wrapper


// HashSet
// Uses hashCode and hashTable(array) (table[obj.hashCode() % table.lenngth])
// Uses linked list in case of collisions
// If there are a lot of collisions, uses RB-tree, so the complexity is O(logN)


// TreeSet<E> implements NavigableSet<E>
// You can get first, last elements
// First element that is greater / lower then specified element
// All elements that greater / less then specified
// ...


// LIST<E>
public interface List<E> extends Collection<E> {
	boolean addAll(int index, Collection<? extends E> c);
	void sort(Comparator<? super E> c);
	E get(int index);
	E set(int index, E element);
	void add(int index, E element);
	E remove(int index); // there is remove(Object o) for remove element
	int indexOf(Object o);
	int lastIndexOf(Objectt o);
	ListIterator<E> listIterator(); // extention of iterator, you can go forward,
	                                // backward. Add and remove element
	ListIterator<E> listIterator(int index);
	List<E> subList(int fromIndex, int toIndex); // returns view, not a copy,
	// if you change element in origin list, sublist also has changed
}

// Example, RangeList
public static List<Integer> rangeList(int fromInclusive, int toExclusive) {
	return new AbstractList<>() {
		@Override
		public Integer get(int index) {
			return fromInclusive + index;
		}
		@Override
		public int size() { return toExclusive - fromInclusive; }
	};
}

// Note, many standard methods have to implementations for lists with random
// access and without. To say, that you list has random access add interface
// RandomAccess. There are not any methods in it.
class RangeList extends AbstractList<Integer> implements RandomAccess {
	...
}

// Standard lists
// ArrayList - standard changable list
// Arrays.asList - changable wrapper under list. You can change elements
//                 but not add / remove elements
// Collections.emptyList() - unchangable empty list
// Collections.singletonList(x) - unchangable list with one element
// Collections.nCopies(n, x) - unchangable list with n identical elements
// List.of(a, b, c) - unchangable list with passed elements
// List.copyOf(list) - unchagable copy on list
// Collections.unmodifiableList(list) - unchangable wrapper under list
// Collections.synchronizedList(list) - synchronized wrapper under list
// Collections.checkedList(list, type)



// QUEUE
// Queue<E> extends Collection<E>
// Methods to add, remove and look at first element. First is safe, returns 
// null/false, if can't execute the operation. Second throws exceptions in this case.
// offer/add - to add an element
// poll/remove - to remove an element
// peek/element - to return first element in collection without removing



// DEQUE
// Deque<E> extends Queue<E>
// With deque you can add / remove elements from both ends
// Also first method is safe, second throws exception in case of failure.
// offerFirst/addFirst
// offerLast/addLast
// pollFirst/removeFirst
// pollLast/removeLast
// peekFirst/getFirst
// peekLast/getLast

// Note, Deque<E> extends Queue<E> therefore it has methods add, remove and element
// They manupulate with elements at the end of the deque.
// But highly recommended to use First/Last methods not to get confused 



// STANDARD QUEUE AND DEQUE
// ArrayDeque - use as deque, queue and stack
// PriorityQueue - priority quese

// There are several old collection, it's better don't use them, use more new once.
// Enumeration -> Iterator
// Vector -> ArrayList
// Stack -> ArrayDeque
// Dictionary -> Map
// Hashtable -> HashMap
// LinkedList -> ArrayList / ArrayDeque - there are memory and time overheads compared 
// to ArrayList. Don't use it.



// MAP
// Map doesn't implement Iterable<E> and Collection<E>. It has its own hierachy.
// There are interfaces SortedMap and NavigableMap.
// We usually use NavigableMap, Sorted is used seldom.
// Map is a collection of Entries - pairs key - value
interface Map<K, V> {
	int size();
	boolean isEmpty();
	// containsKey usually works fast O(1) or O(NlogN)
	boolean containsKey(Object key); //Note, Object not K, it wrong solution
	// containsValue usually works slow O(n)
	boolean containsValue(Object value); //Note, Object not K, it wrong solution
	V get(Object key); // return null, there is no this key
	V put(K key, V value); // add new value
	V remove(Object key);
	void putAll(Map<? extends K, ? extends V> m);
	void clear();
	Set<K> keySet(); // all keys, return views
	Collection<V> values(); // all values, values can be not unique, return view
	Set<Map,Entry<K, V>> entrySet(); // all entries key-value, return view
	V getOrDefault(Object key, V value);
	V putIfAbsent(K key, V value); // add pair, if there is no 
	// the key in the map or it is null
	boolean remove(Ojbect key, Object value); // remove on this key-value pair
	boolean replace(K key, V oldValue, V newValue); // replace only specified value
	V replace(K key, V value); // replace any value
}

// Usually entry is a view, so if you will change value, the value in the 
// collection will also be changed`
Interface Entry<K, V> {
	K getKey();
	V getValue();
	V setValue(V value);
}

// Example, if you want go over the map and get key-value pairs, 
// better to use entrySet rather then keySet and get.
// It's because entrySet usually works faster.
public static void iterateOverMap(Map<String, Integer> map) {
	for(Map.Entry<String, Integer> entry: map.entrySet()) {
		System.out.println(entry.getKey() + " -> " + entry.getValue());
	}
}

// Iterate with forEach and BiConsumer interface
public static void iterateOverMap(Map<String, Integer> map) {
	map.forEach(new BiConsumer<String, Integer>() {
		@Override
		public void accept(String key, Integer value) {
			System.out.println(key + " -> " + value);
		}
	});
}
// Iterate with forEach and lambda
// It's the best way, because we should not create iterator
public static void iterateOverMap(Map<String, Integer> map) {
	map.forEach((key, value) -> System.out.println(key + " -> " + value));
}

// Let's implement map with pairs, number and string representation of number
// We will use anonymous type, and should just implement the method entrySet,
// that returns an entrySet
static Map<Integer, String> numberToString(int count) {
	return new AbstractMap<>() {
		@Override
		public Set<Entry<Integer, String>> entrySet() {
			return new AbstractSet<>() {
				@Override
				public Iterator<Entry<Integer, String>> iterator() {
					return new Iterator<>() {
						int next = 0;
						@Override
						public Entry<Integer, String> next() {
							return new SimpleImmutableEntry<>(next, String.valueOf(next++));
						}

						@Override
						public boolean hasNext() {
							return next < count;
						}
					};
				}

				@Override
				public int size() {
					return count;
				}
			};
		}
	};
}


// MODIFICATION
// With keySet, it's the slowest way
public static void trimAllValues(Map<String, String> map) {
	for(String key: map.keySet()) {
		String value = map.get(key);
		map.put(key, value.trim());
	}
}
// With entries 
public static void trimAllValues(Map<String, String> map) {
	for(Map.Entry<String, String> entry: map.entrySet()) {
		String value = entry.getValue();
		entry.setValue(value.trim());
	}
}
// With replaceAll, it's a faster way
public static void trimAllValues(Map<String, String> map) {
	map.replaceAll((key, value) -> value.trim());
}

// REMOVING
// With keys
public static void removeUnwantedValues(Map<String, String> map) {
	for(String key: map.keySet()) {
		String value = map.get(key);
		if(value.equals("aaa") || value.equals("bbb")) {
			map.remove(key);
		}
	}
}
// With entries
public static void removeUnwantedValues(Map<String, String> map) {
	Iterator<Map.Entry<String, String>> iterator = map.entrySet().iterator();
	while(iterator.hasNext()) {
		Map.Entry<String, String> entry = iterator.next();
		if(entry.getValue() == "aaa" || entry.getValue() == "bbb") {
			iterator.remove();
		}
	}
}
// With REMOVEIF method, every collection has it.
public static void removeUnwantedValues(Map<String, String> map) {
	map.entrySet().removeIf(entry -> entry.getValue().equals("aaa") || entry.getValue().equals("bbb"));
}
// With values, if you remove value => you remove all entry
public static void removeUnwantedValues(Map<String, String> map) {
	map.values().removeIf(value -> value.equals("aaa") || value.equals("bbb"));
}
// With remove all method. If you add value for value view, you remove 
// all entry
public static void removeUnwantedValues(Map<String, String> map) {
	List<String> unwantedValues = Arrays.asList("aaa", "bbb");
	map.values().removeAll(unwantedValues);
}



// MULTIMAP
// Sometimes, you want to store several value that correspont one key
// It is named multimap
static Map<String, List<String>> map = new HashMap<>();
public static void add(String key, String value) {
	List<String> list = map.get(key);
	if(list == null) {
		list = new ArrayList<>();
		map.put(key, list);
	}
	list.add(value);
}
// You can use COMPUTEIFABSENT to add value, if key is not present. It
// returns value(created or not)
public static void add(String key, String value) {
	map.computeIfAbsent(key, k -> new ArrayList<>()).add(value);
}



// MULTI-SET/BAG
// We want to count element occurances
static Map<String, Integer> counts = new HashMap<>();
public static void add(String key) {
	Integer count = counts.get(key);
	if(count == null) {
		count = 0;
	}
	counts.put(key, count + 1);
}
// We can use the method MERGE, it accepts key, value and function that
// return new value if there is oldValue. It accepts oldValue and passed value
public static void add(String key) {
	counts.merge(key, 1, (a, b) -> a + b);
}



// STANDARD MAPS
// HashMap - unordered changable standard map
// LinkedHashMap - ordered changable standard map. It stores elements in order
// of thier addition to the map
// TreeMap - sorted changable Map
// EnumMap - a map, which key are elements of the enum
// IdentityHashMap - map, which keys are compared with ==, not equals
// Collections.singletonMap(k, v) - unchangable map with one element
// Collections.emptyMap() - unchangable empty map
// Map.of(k1, v1, k2, v2, ...); Map.ofEntries(Map.entry(k1, v1), ...) -
// unchangable map
// Collections.unmodifiableMap, synchronizedMap, checkedMap - wrappers



// NAVIGABLEMAP
// NavigableMap<K, V> extends SortedMap<K, V>
// For example, you can find greatest key/entry, that lower then specified 
// get view that started from specified key, revert map(descending map)
// - lowerKey, lowerEntry, floorKey, floorEntry
// - higerKey, higherEntry, ceilingKey, ceilingEntry
// - firstKey, firstEntry, lastKey, lastEntry
// - pollFirstEntry, pollLastEntry
// - descendingMap, navigableKeySet, descendingKeySet
// - subMap, headMap, tailMap



// CONCURRENTMODIFICATIONEXCEPTION
// - Is thrown if collection has noticed that it is changed, when you can'table
//   do this
// - Usually while iteration with iterator or with forEach
// - You can modify a collection while iterating only with with iterator. 
 


